docs/LoaderApplicationInterface.md

```diff
@@ -0,0 +1,686 @@
+
+An even better solution would be for an application to perform a
+`vkGetDeviceProcAddr` call on all device functions.
+This further optimizes the call chain by removing the loader all-together under
+most scenarios:
+
+![Application Device Call Chain](./images/loader_device_chain_app.png)
+
+Also, notice if no layers are enabled, the application function pointers point
+**directly to the implementation**.
+With many function calls, the lack of indirection in each adds up to non-trivial
+performance savings.
+
+**NOTE:** There are some device functions which still require the loader to
+intercept them with a *trampoline* and *terminator*.
+There are very few of these, but they are typically functions which the loader
+wraps with its own data.
+In those cases, even the device call chain will continue to look like the
+instance call chain.
+One example of a device function requiring a *terminator* is
+`vkCreateSwapchainKHR`.
+For that function, the loader needs to potentially convert the KHR_surface
+object into an implementation-specific KHR_surface object prior to passing down
+the rest of the function's information to the implementation.
+
+Remember:
+ * `vkGetInstanceProcAddr` is used to query instance and physical device
+   functions, but can query all functions.
+ * `vkGetDeviceProcAddr` is only used to query device functions.
+
+
+### ABI Versioning
+
+The Vulkan loader library will be distributed in various ways including Vulkan
+SDKs, OS package distributions and Independent Hardware Vendor (IHV) driver
+packages.
+These details are beyond the scope of this document.
+However, the name and versioning of the Vulkan loader library is specified so
+an app can link to the correct Vulkan ABI library version.
+ABI backwards compatibility is guaranteed for all versions with the same major
+number (e.g. 1.0 and 1.1).
+On Windows, the loader library encodes the ABI version in its name such that
+multiple ABI incompatible versions of the loader can peacefully coexist on a
+given system.
+The Vulkan loader library file name is `vulkan-<ABI version>.dll`.
+For example, for Vulkan version 1.X on Windows the library filename is
+`vulkan-1.dll`.
+This library file can typically be found in the `windows\system32`
+directory (on 64-bit Windows installs, the 32-bit version of the loader with
+the same name can be found in the `windows\sysWOW64` directory).
+
+For Linux and macOS, shared libraries are versioned based on a suffix.
+Thus, the ABI number is not encoded in the base of the library filename as on
+Windows.
+When an application or a larger library links to `libvulkan.so` (with something
+like "cc -lvulkan"), the SONAME of the library (currently `libvulkan.so.1`) is
+what gets hard-coded into the binary, so it will always load the same major
+version that it was compiled against.
+On Linux an application wanting to link to the latest Vulkan ABI version would
+just link to the name vulkan (`libvulkan.so`).
```

Comment:
Sorry, I hadn't realised you were going to incorporate my comment into the document as-is - I was writing review comments in a conversational style, rather than writing for inclusion in documentation.

I think the current text  has the potential to be misleading, because there are two ways an application can link to the Vulkan library - ordinary compile-time linking, and dynamic loading at runtime - and they have different patterns.

For compile-time linking, like `vkcube` does, what we want (and the only thing that works in most build systems) is for the application to ask for `-lvulkan`, and the compiler and linker to resolve this to a specific SONAME `libvulkan.so.1`.

For dynamic loading, like SDL does, what we want is for the developer to load `libvulkan.so.1` specifically.

How about this, perhaps?

```suggestion
On Linux, applications that have a hard dependency on Vulkan should request
linking to the unversioned name `libvulkan.so` in their build system, for
example by importing the CMake target `Vulkan::Vulkan` or by using the
output of `pkg-config --cflags --libs vulkan` as compiler flags.
As usual for Linux libraries, the compiler and linker will resolve this to
a dependency on the correct versioned SONAME, currently `libvulkan.so.1`.
Linux applications that load Vulkan-Loader dynamically at runtime do not
benefit from this mechanism, and should instead make sure to pass the
versioned name such as `libvulkan.so.1` to `dlopen()`, to ensure that they
load a compatible version.
```

Technically this is also true for most non-Linux ELF systems like FreeBSD and Solaris, but I know Linux is the only one that's officially supported, and developers on non-Linux Unix are probably already familiar with extrapolating what's meant to happen on their platform from what happens on Linux.

I don't know how this works on macOS (I don't develop on Apple systems).

Reply:
I like your version.  Plus, I ran it by the internal Mac savy people and they say the language works there as well.